arangod/Scheduler/LockfreeThreadPool.cpp

```diff
@@ -0,0 +1,133 @@
+/// DISCLAIMER
+///
+/// Copyright 2014-2024 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Business Source License 1.1 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+#include <algorithm>
+#include <atomic>
+#include <memory>
+#include <thread>
+
+#ifdef TRI_HAVE_SYS_PRCTL_H
+#include <sys/prctl.h>
+#endif
+
+#include "Basics/cpu-relax.h"
+#include "Logger/LogMacros.h"
+
+#include "LockfreeThreadPool.h"
+
+namespace arangodb {
+
+namespace {
+struct StopWorkItem : LockfreeThreadPool::WorkItem {
+  void invoke() noexcept override {}
+} stopItem;
+
+}  // namespace
+
+LockfreeThreadPool::LockfreeThreadPool(const char* name,
+                                       std::size_t threadCount,
+                                       ThreadPoolMetrics metrics)
+    : numThreads(threadCount), _threads(), _metrics(metrics) {
+  unsigned cnt = 1;
+  std::generate_n(std::back_inserter(_threads), threadCount, [&]() {
+    auto id = cnt++;
+    auto thread = std::jthread([this, id]() noexcept {
+      while (true) {
+        auto item = pop(id);
+        if (item.get() == &stopItem) {
+          std::ignore = item.release();
+          return;
+        }
+
+        try {
+          item->invoke();
+        } catch (...) {
+          LOG_TOPIC("d5fb4", WARN, Logger::FIXME)
+              << "Scheduler just swallowed an exception.";
+        }
+        statistics.done += 1;
+      }
+    });
+
+#ifdef TRI_HAVE_SYS_PRCTL_H
+    pthread_setname_np(thread.native_handle(), name);
+#endif
+
+    return thread;
+  });
+}
+
+LockfreeThreadPool::~LockfreeThreadPool() { shutdown(); }
+
+void LockfreeThreadPool::shutdown() noexcept {
+  // push as many stop items as we have threads
+  for ([[maybe_unused]] auto& thread : _threads) {
+    push(std::unique_ptr<WorkItem>(&stopItem));
+  }
+  for (auto& thread : _threads) {
+    if (thread.joinable()) {
+      thread.join();
+    }
+  }
+  // pop remaining items from the queue and release them
+  WorkItem* item;
+  while (_queue.pop(item)) {
+    if (item != &stopItem) {
+      delete item;
+    }
+  }
+}
+
+void LockfreeThreadPool::push(std::unique_ptr<WorkItem>&& task) noexcept {
+  _queue.push(task.release());
+  statistics.queued.fetch_add(1);
+  auto old = statistics.inQueue.fetch_add(1);
```

Comment:
Are we intentionally using the strictest memory order here, or can it be relaxed?

Reply:
The LockfreeThreadPool turned out to scale less well so I did not spend time on such low level optimizations - it will probably be removed at some point.